src/core/lib/security/transport/client_auth_filter.cc

```diff
@@ -279,7 +280,7 @@ static void auth_start_transport_stream_op_batch(
     GPR_ASSERT(batch->payload->context != nullptr);
     if (batch->payload->context[GRPC_CONTEXT_SECURITY].value == nullptr) {
       batch->payload->context[GRPC_CONTEXT_SECURITY].value =
```

Comment:
It's not clear to me why this is being set in `start_transport_stream_op_batch()` in the first place.  Is there any reason we can't move this to `init_call_elem()`?  If we did that, there would be no need to store the arena pointer in the call data.

Reply:
Oops I thought this sent yesterday evening but I forgot the bus wifi wasn't working...here's my response.

I'm not familiar enough with the logic of the code to know the reason, but just from code inspection, it looks like whether or not this is set is being used to decide what code path to take later on. 

start_transport_stream_op_batch calls on_host_checked if we're sending initial metadata. At this point, batch->cancel_stream may or may not be true, so batch->payload->context[GRPC_CONTEXT_SECURITY].value could be either nullptr or not nullptr. 
on_host_checked calls send_security_metadata, which checks if context[GRPC_CONTEXT_SECURITY].value is nullptr and goes down a different code path depending on that check. In particular, it skips sending the metadata altogether and just calls the next filter with grpc_call_next_op. 

Questions: 1) Can we set the batch's payload in init_call_elem? 2) In what cases do we want the context[GRPC_CONTEXT_SECURITY].value to be nullptr and what does it signify? And can / should the code concerning this be reorganized? The person who can probably best answer question 2 is on vacation right now.

Reply:
We can't set the batch's payload in `init_call_elem()`, but I don't think we need to.  We can set the call's context there, which is available in `args->context`.

I don't think the current code is ever calling `send_security_metadata()` in the case where the context value has not ever been set.  The reason I say that is that the the current code sets the context value whenever the filter sees a batch that does not contain a `cancel_stream` op and the context value is not already set.  However, a `cancel_stream` op is always started in its own batch (never in the same batch with other ops).  And in order to get to `send_security_metadata()`, we have to have seen a batch with a `send_initial_metadata` op.  Since we know that batch will not have included a `cancel_stream` op, I think the context value will always be set by the time we get to `send_security_metadata()`.

I think it should be safe to move the code that sets the context into `init_call_elem()`.  But we should get confirmation from the security folks.  If @jiangtaoli2016 is on vacation, ask @jboeuf (the original author of this code).

Reply:
Thanks for the response, Mark. I wasn't aware of the semantics around cancel_stream ops. 

@jiangtaoli2016 , are you okay with moving the code that sets the security context into init_call_elem, instead of having it in start_transport_stream_op_batch?

Reply:
@markdroth It looks like we need to access the `client_security_context` in `send_security_metadata()`, so we'll need to access it in `auth_start_transport_stream_op_batch`. Could we set the call's context in `init_call_elem()` and store it in `args->context` and then put a pointer to the context in the call data? Or is there some other way to access args->context from within `start_transport_stream_op_batch`?

Reply:
As per our discussion, the call context is available as both `args->context` in `init_call_elem()` and `batch->payload->context` in `start_transport_stream_op_batch()`, so the way you've changed this looks good.